src/coreclr/vm/jitinterface.cpp

```diff
@@ -10623,10 +10631,41 @@ void* CEEJitInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,         /* IN  */
             dynamicFtnNum == DYNAMIC_CORINFO_HELP_CHKCASTINTERFACE ||
             dynamicFtnNum == DYNAMIC_CORINFO_HELP_CHKCASTCLASS ||
             dynamicFtnNum == DYNAMIC_CORINFO_HELP_CHKCASTCLASS_SPECIAL ||
-            dynamicFtnNum == DYNAMIC_CORINFO_HELP_UNBOX)
+            dynamicFtnNum == DYNAMIC_CORINFO_HELP_UNBOX ||
+            dynamicFtnNum == DYNAMIC_CORINFO_HELP_ARRADDR_ST ||
+            dynamicFtnNum == DYNAMIC_CORINFO_HELP_LDELEMA_REF)
         {
             Precode* pPrecode = Precode::GetPrecodeFromEntryPoint((PCODE)hlpDynamicFuncTable[dynamicFtnNum].pfnHelper);
             _ASSERTE(pPrecode->GetType() == PRECODE_FIXUP);
+            
+            MethodDesc* helperMD = pPrecode->GetMethodDesc();
+            _ASSERT(helperMD != nullptr);
+
+            // Check if the target MethodDesc is already jitted to its final Tier
+            // so we no longer need to use indirections and can emit a direct call instead.
+            NativeCodeVersion::OptimizationTier tier;
+
+            // Avoid taking the lock for foreground jit compilations
+            if (TieredCompilationManager::IsBackgroundWorkerActive())
```

Comment:
Are all of these methods eligible for tiering, or does it need to be checked? Maybe an assert would do?

Reply:
Yes, all of them point to managed calls in CastHelpers.* so can benefit from being a direct call instead of indirect.